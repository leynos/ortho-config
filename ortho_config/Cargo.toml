[package]
name = "ortho_config"
version = "0.2.0"
edition = "2024"

[lints]
workspace = true

[dependencies]
ortho_config_macros = { path = "../ortho_config_macros" }
serde = { version = "1", features = ["derive"] }
thiserror = "1"
clap = { version = "4", features = ["derive"] }
clap-dispatch = "0.1"
figment = { version = "0.10", default-features = false, features = ["env", "test"] }
uncased = "0.9"
xdg = "3"
toml = { version = "0.8", optional = true }
figment-json5 = { version = "0.1", optional = true }
serde_yaml = { version = "0.9", optional = true }

[features]
default = ["toml"]
toml = ["figment/toml", "dep:toml"]
json5 = ["dep:figment-json5"]
yaml = ["figment/yaml", "dep:serde_yaml"]

[dev-dependencies]
figment = { version = "0.10", features = ["toml", "env"] }
serde = { version = "1", features = ["derive"] }
trybuild = "1"
