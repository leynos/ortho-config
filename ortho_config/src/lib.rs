//! Core crate for the `OrthoConfig` configuration framework.
//!
//! This crate defines the [`OrthoConfig`] trait and supporting error types. The
//! actual implementation of the derive macro lives in the companion
//! `ortho_config_macros` crate.

pub use ortho_config_macros::OrthoConfig;

mod error;
mod file;

pub use error::OrthoError;
pub use file::load_config_file;

/// Trait implemented for structs that represent application configuration.
pub trait OrthoConfig: Sized + serde::de::DeserializeOwned {
    /// Loads, merges, and deserializes configuration from all available
    /// sources according to predefined precedence rules.
    #[allow(clippy::result_large_err)]
    fn load() -> Result<Self, OrthoError>;
}
