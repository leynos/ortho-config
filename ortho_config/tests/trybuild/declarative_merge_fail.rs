use ortho_config::OrthoConfig;
use serde::Deserialize;

#[derive(Debug, Deserialize, OrthoConfig)]
struct UnsupportedStrategy {
    #[serde(default)]
    #[ortho_config(merge_strategy = "unknown")]
    values: Vec<String>,
}

fn main() {}
