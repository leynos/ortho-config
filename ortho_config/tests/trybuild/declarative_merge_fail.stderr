error: unknown merge_strategy 'unknown'; expected one of "append", "replace", or "keyed"
 --> tests/trybuild/declarative_merge_fail.rs:7:37
  |
7 |     #[ortho_config(merge_strategy = "unknown")]
  |                                     ^^^^^^^^^
